shards/modules/network/network.cpp

```diff
@@ -546,16 +564,58 @@ struct Server : public NetworkBase {
           context->main->dispatcher.trigger(std::move(event));
         }
 
-        auto nextSize = ikcp_peeksize(peer->kcp);
-        while (nextSize > 0) {
-          _buffer.resize(nextSize);
-          auto size = ikcp_recv(peer->kcp, (char *)_buffer.data(), nextSize);
-          assert(size == nextSize);
+        bool isMessageReady = false;
+
+        {
+          std::scoped_lock peerLock(peer->peerMutex);
 
+          peer->maybeUpdate();
+
+          // Initialize variables for buffer management.
+          size_t offset = peer->recvBuffer.size();
+          auto nextChunkSize = ikcp_peeksize(peer->kcp);
+
+          // Loop to receive all available chunks.
+          while (nextChunkSize > 0) {
+            // SHLOG_TRACE("Received chunk of size: {}, offset: {}", nextChunkSize, offset);
+
+            // Resize the buffer to hold the incoming chunk.
+            peer->recvBuffer.resize(nextChunkSize + offset);
+
+            // Receive the chunk.
+            auto receivedSize = ikcp_recv(peer->kcp, (char *)peer->recvBuffer.data() + offset, nextChunkSize);
+            assert(receivedSize == nextChunkSize);
+
+            // If this is the first chunk, read the expected total size from its prefix.
+            if (offset == 0) {
+              peer->expectedSize = *(uint32_t *)peer->recvBuffer.data();
+              // if (peer->expectedSize > 25000) {
+              //   SHLOG_TRACE("Receiving big message: {}", peer->expectedSize);
+              // }
+            }
+
```

Comment:
Can you put this in a function and use it in the code below, looks like it does the same?